src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheck.java

```diff
@@ -297,44 +294,48 @@ public int[] getRequiredTokens() {
     @Override
     public void beginTree(DetailAST rootAST) {
         matcher = format.matcher(getFileContents().getText().getFullText());
-        matchCount = 0;
         errorCount = 0;
         findMatch();
     }
 
     /**
-     * Recursive method that finds the matches.
+     * Finds the matches in the file.
      *
-     * @noinspection TailRecursion
-     * @noinspectionreason TailRecursion - until issue #14814
      */
     // suppress deprecation until https://github.com/checkstyle/checkstyle/issues/11166
     @SuppressWarnings("deprecation")
     private void findMatch() {
-        final boolean foundMatch = matcher.find();
-        if (foundMatch) {
-            final FileText text = getFileContents().getText();
-            final LineColumn start = text.lineColumn(matcher.start());
-            final int startLine = start.getLine();
-
-            final boolean ignore = isIgnore(startLine, text, start);
-
-            if (!ignore) {
-                matchCount++;
-                if (illegalPattern || checkForDuplicates
-                        && matchCount - 1 > duplicateLimit) {
-                    errorCount++;
-                    logMessage(startLine);
+        boolean valid = true;
+        int matchCount = 0;
+        while (valid) {
```

Comment:
Could you simplify it by using a different condition? Something like `while (matcher.find() && errorCount < errorLimit)`. This will require more analysis and understanding of the current logic, and we probably won't need `canContinueValidation`.

Reply:
done